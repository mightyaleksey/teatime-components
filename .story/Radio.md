## Preview


### size s

```
<div>
  button theme
  <Radio
    name='motorrad-s'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='s'
    theme='button'/>
  {' '}
  <Radio
    disabled={true}
    name='motorrad-ds'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='s'
    theme='button'/>
</div>
```

```
<div>
  common theme
  <Radio
    name='motorrad-cs'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='s'
    theme='common'/>
  {' '}
  <Radio
    disabled={true}
    name='motorrad-cds'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='s'
    theme='common'/>
</div>
```


### size m

```
<div>
  button theme
  <Radio
    name='motorrad-m'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='m'
    theme='button'/>
  {' '}
  <Radio
    disabled={true}
    name='motorrad-dm'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='m'
    theme='button'/>
</div>
```

```
<div>
  common theme
  <Radio
    name='motorrad-cm'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='m'
    theme='common'/>
  {' '}
  <Radio
    disabled={true}
    name='motorrad-cdm'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='m'
    theme='common'/>
</div>
```


### size l

```
<div>
  button theme
  <Radio
    name='tumbler-l'
    options={[
      {label: 'On', value: 'on'},
      {label: 'Off', value: 'off'},
    ]}
    size='l'
    theme='button'/>
  {' '}
  <Radio
    defaultValue='on'
    disabled={true}
    name='tumbler-dl'
    options={[
      {label: 'On', value: 'on'},
      {label: 'Off', value: 'off'},
    ]}
    size='l'
    theme='button'/>
</div>
```

```
<div>
  common theme
  <Radio
    name='motorrad-cl'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='l'
    theme='common'/>
  {' '}
  <Radio
    disabled={true}
    name='motorrad-cdl'
    options={[
      {label: 'Yamaha', value: 'yamaha'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'MZ', value: 'mz'},
      {label: 'Kawasaki', value: 'kawasaki'},
    ]}
    size='l'
    theme='common'/>
</div>
```
