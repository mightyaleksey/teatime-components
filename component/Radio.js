'use strict';

const {Component, PropTypes} = require('react');
const {filterProps, genericName, isControlled} = require('../lib/util');
const {genericId} = require('../lib/identity');
const {isUndefined, map, noop} = require('../lib/dash');
const Box = require('../view/Box');
const React = require('react');
const cc = require('classnames');
const warn = require('../lib/warn');

const cssModules = {
  'button-l': require('../style/radio/radio-button-l.css'),
  'button-m': require('../style/radio/radio-button-m.css'),
  'button-s': require('../style/radio/radio-button-s.css'),
  'common-l': require('../style/radio/radio-common-l.css'),
  'common-m': require('../style/radio/radio-common-m.css'),
  'common-s': require('../style/radio/radio-common-s.css'),
};

let didWarnForRadioDefaultValue = false;

class Radio extends Component {
  constructor(props) {
    super(props);

    this._controlled = isControlled(props);

    if (process.env.NODE_ENV !== 'production')
      if (this._controlled && !isUndefined(props.defaultValue) && !didWarnForRadioDefaultValue) {
        didWarnForRadioDefaultValue = true;
        warn(true, 'defaultValue', 'Radio', 'radio');
      }

    const value = this._controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      prefix: genericId(),
      value: isUndefined(value) ? null : value,
    };
  }

  componentWillReceiveProps(nextProps) {
    this._controlled = isControlled(nextProps);

    this.setState({
      value: this._controlled
        ? nextProps.value
        : this.state.value,
    });
  }

  css = tokenName => genericName(this.props, tokenName)

  focus = noop

  _onChange = (e, {value}) => {
    if (!this._controlled) this.setState({value});
    this.props.onChange(e, {value});
  }

  computeOptions(options) {
    const {disabled, name, theme} = this.props;
    const {prefix, value} = this.state;

    const {css} = this;
    const styles = css();

    const hasLabel = theme === 'common';

    return map(option =>
      this.renderOption(option, {
        checked: option.value === value,
        disabled: option.disabled || disabled,
        hasLabel,
        key: `${prefix}${option.value}`,
        name,
        onChange: this._onChange,
        styles,
        type: 'radio',
      }), options);
  }

  render() {
    const {className, id, options} = this.props;
    const {css} = this;

    return (
      <div
        {...filterProps(this.props)}
        className={cc(css('container'), className)}
        id={id}
        onChange={void 0}>
        {this.computeOptions(options)}
      </div>
    );
  }

  renderOption(optionProps, genericProps) {
    return (
      <Box
        {...optionProps}
        {...genericProps}/>
    );
  }
}

Radio.defaultProps = {
  onChange: noop,
  size: 'm',
  styles: cssModules,
  theme: 'common',
};

Radio.propTypes = {
  className: PropTypes.string,
  defaultValue: PropTypes.any,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.object).isRequired,
  size: PropTypes.oneOf([
    'l',
    'm',
    's',
  ]),
  styles: PropTypes.object,
  theme: PropTypes.oneOf([
    'button',
    'common',
  ]),
  value: PropTypes.any,
};

module.exports = Radio;
