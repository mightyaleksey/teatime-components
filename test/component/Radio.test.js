import Radio from '../../component/Radio';
import React from 'react';
import renderer from 'react-test-renderer';

test('Radio renders correctly with defaults', () => {
  const tree = renderer.create(
    <Radio
      name='radio control'
      options={[
        {label: 'Ducati', value: 'ducati'},
        {label: 'Honda', value: 'honda'},
        {label: 'Jawa', value: 'jawa'},
      ]}/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Radio theme=`button` renders correctly', () => {
  const tree = renderer.create(
    <Radio
      className='outer'
      id='radios id'
      name='radio control'
      options={[
        {label: 'Ducati', value: 'ducati'},
        {disabled: true, label: 'Honda', value: 'honda'},
        {label: 'Jawa', value: 'jawa'},
      ]}
      theme='button'
      value='ducati'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Radio renders correctly with disabled state', () => {
  const tree = renderer.create(
    <Radio
      className='outer'
      disabled={true}
      id='radios id'
      name='radio control'
      options={[
        {label: 'Ducati', value: 'ducati'},
        {label: 'Honda', value: 'honda'},
        {label: 'Jawa', value: 'jawa'},
      ]}
      theme='common'
      value='ducati'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
