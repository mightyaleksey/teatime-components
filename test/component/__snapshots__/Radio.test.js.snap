// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly with defaults 1`] = `
<div
  className="container"
  id={undefined}
  onChange={undefined}
>
  <span
    className="wrapper"
  >
    <input
      checked={false}
      className="native"
      defaultValue={undefined}
      disabled={undefined}
      id="_teatime2"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="ducati"
    />
    <label
      className="control"
      htmlFor="_teatime2"
    />
    <label
      className="label"
      htmlFor="_teatime2"
    >
      Ducati
    </label>
  </span>
  <span
    className="wrapper"
  >
    <input
      checked={false}
      className="native"
      defaultValue={undefined}
      disabled={undefined}
      id="_teatime3"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="honda"
    />
    <label
      className="control"
      htmlFor="_teatime3"
    />
    <label
      className="label"
      htmlFor="_teatime3"
    >
      Honda
    </label>
  </span>
  <span
    className="wrapper"
  >
    <input
      checked={false}
      className="native"
      defaultValue={undefined}
      disabled={undefined}
      id="_teatime4"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="jawa"
    />
    <label
      className="control"
      htmlFor="_teatime4"
    />
    <label
      className="label"
      htmlFor="_teatime4"
    >
      Jawa
    </label>
  </span>
</div>
`;

exports[`Radio renders correctly with disabled state 1`] = `
<div
  className="container outer"
  id="radios id"
  onChange={undefined}
>
  <span
    className="wrapper"
  >
    <input
      checked={true}
      className="native"
      defaultValue={undefined}
      disabled={true}
      id="_teatime10"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="ducati"
    />
    <label
      className="control"
      htmlFor="_teatime10"
    />
    <label
      className="label"
      htmlFor="_teatime10"
    >
      Ducati
    </label>
  </span>
  <span
    className="wrapper"
  >
    <input
      checked={false}
      className="native"
      defaultValue={undefined}
      disabled={true}
      id="_teatime11"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="honda"
    />
    <label
      className="control"
      htmlFor="_teatime11"
    />
    <label
      className="label"
      htmlFor="_teatime11"
    >
      Honda
    </label>
  </span>
  <span
    className="wrapper"
  >
    <input
      checked={false}
      className="native"
      defaultValue={undefined}
      disabled={true}
      id="_teatime12"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="jawa"
    />
    <label
      className="control"
      htmlFor="_teatime12"
    />
    <label
      className="label"
      htmlFor="_teatime12"
    >
      Jawa
    </label>
  </span>
</div>
`;

exports[`Radio theme=\`button\` renders correctly 1`] = `
<div
  className="container outer"
  id="radios id"
  onChange={undefined}
>
  <span
    className="wrapper"
  >
    <input
      checked={true}
      className="native"
      defaultValue={undefined}
      disabled={undefined}
      id="_teatime6"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="ducati"
    />
    <label
      className="control"
      htmlFor="_teatime6"
    >
      Ducati
    </label>
  </span>
  <span
    className="wrapper"
  >
    <input
      checked={false}
      className="native"
      defaultValue={undefined}
      disabled={true}
      id="_teatime7"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="honda"
    />
    <label
      className="control"
      htmlFor="_teatime7"
    >
      Honda
    </label>
  </span>
  <span
    className="wrapper"
  >
    <input
      checked={false}
      className="native"
      defaultValue={undefined}
      disabled={undefined}
      id="_teatime8"
      name="radio control"
      onChange={[Function]}
      type="radio"
      value="jawa"
    />
    <label
      className="control"
      htmlFor="_teatime8"
    >
      Jawa
    </label>
  </span>
</div>
`;
